text/0000-fn-lifetime.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: fn lifetime
+- Start Date: 2017-01-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a `'fn` lifetime that is bound to the scope of the body of the current
+innermost function or closure.
+
+# Motivation
+[motivation]: #motivation
+
+Doing this will enable us to declare that some values live longer than their
+containing scope, allowing us to "allocate" recursive data structures (e.g.
+linked lists) on the stack, as in:
+
+```rust
+struct Node<T> {
+    data: T,
+    next: Option<&'fn Node>,
```

Comment:
Uhm, this can't possibly work, you need to make it a parameter of `Node` otherwise you end up with C++-level handwavey rules that are (approximately) checked through invasive heuristics.
Either that or the compiler injects the parametricity but that's much harder to do and get right than the existing lifetime system that we already have.